include/wil/safearrays.h

```diff
@@ -0,0 +1,889 @@
+        template<> struct infer_safearray_type_traits<CURRENCY>                   { static constexpr auto vartype = VT_CY; };
+        template<> struct infer_safearray_type_traits<DECIMAL>                    { static constexpr auto vartype = VT_DECIMAL; };
+        template<> struct infer_safearray_type_traits<BSTR>                       { static constexpr auto vartype = VT_BSTR; };
+        template<> struct infer_safearray_type_traits<LPUNKNOWN>                  { static constexpr auto vartype = VT_UNKNOWN; };
+        template<> struct infer_safearray_type_traits<LPDISPATCH>                 { static constexpr auto vartype = VT_DISPATCH; };
+        template<> struct infer_safearray_type_traits<VARIANT>                    { static constexpr auto vartype = VT_VARIANT; };
+
+        inline void __stdcall SafeArrayDestroy(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            FAIL_FAST_IF_FAILED(::SafeArrayDestroy(psa));
+        }
+
+        inline void __stdcall SafeArrayLock(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            FAIL_FAST_IF_FAILED(::SafeArrayLock(psa));
+        }
+
+        inline void __stdcall SafeArrayUnlock(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            FAIL_FAST_IF_FAILED(::SafeArrayUnlock(psa));
+        }
+
+        template<typename T>
+        inline void __stdcall SafeArrayAccessData(SAFEARRAY* psa, T*& p) WI_NOEXCEPT
+        {
+            FAIL_FAST_IF_FAILED(::SafeArrayAccessData(psa, reinterpret_cast<void**>(&p)));
+        }
+
+        inline void __stdcall SafeArrayUnaccessData(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            FAIL_FAST_IF_FAILED(::SafeArrayUnaccessData(psa));
+        }
+
+        inline VARTYPE __stdcall SafeArrayGetVartype(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            VARTYPE vt = VT_NULL;   // Invalid for SAs, so use to mean SA was null
+            if ((psa != nullptr) && FAILED(::SafeArrayGetVartype(psa, &vt)))
+            {
+                vt = VT_EMPTY;  // Invalid for SAs, use to mean type couldn't be determined
+            }
+            return vt;
+        }
+
+        inline ULONG __stdcall SafeArrayGetLockCount(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            return (psa != nullptr) ? psa->cLocks : 0U;
+        }
+
+        inline HRESULT __stdcall SafeArrayCreate(VARTYPE vt, UINT dims, SAFEARRAYBOUND* sab, SAFEARRAY*& psa) WI_NOEXCEPT
+        {
+            WI_ASSERT(sab != nullptr);
+            WI_ASSERT(dims > 0);
+            psa = ::SafeArrayCreate(vt, dims, sab);
+            RETURN_LAST_ERROR_IF_NULL(psa);
+            WI_ASSERT(vt == details::SafeArrayGetVartype(psa));
+            return S_OK;
+        }
+
+        inline HRESULT __stdcall SafeArrayDimElementCount(SAFEARRAY* psa, UINT dims, ULONG* result) WI_NOEXCEPT
+        {
+            __FAIL_FAST_ASSERT__(psa != nullptr);
+            __FAIL_FAST_ASSERT__(result != nullptr);

```

Comment:
Both of these asserts are unnecessary; you dereference both below. In general, asserting against null is discouraged

Reply:
Yeah, I don't see a lot of other nullptr asserts in WIL, so I'll take them out.